src/macaron/slsa_analyzer/git_url.py

```diff
@@ -106,76 +107,59 @@ def check_out_repo_target(git_obj: Git, branch_name: str = "", digest: str = "",
-
-    # We only pull the latest changes if we are not running in offline mode and:
-    #   - no digest is provided.
-    #   - or a commit digest is provided but it does not exist in the current local branch.
-    if not offline_mode and (not digest or not commit_exists(git_obj, digest)):
-        logger.info("Pulling the latest changes of branch %s fast-forward only.", res_branch)
-        if not pull_latest_changes(git_obj):
-            logger.error(
-                "Cannot pull the latest changes for branch %s.",
-                res_branch,
-            )
+            logger.error("Unable to fetch from the remote repository. Error: %s", error)
             return False
 
-    if digest:
-        current_head: Commit = git_obj.repo.head.commit
-        # We don't return False if current_head is None here to try checking out the
-        # specific commit by the user.
-        if current_head and current_head.hexsha == digest:
-            logger.info("HEAD of the repo is already at %s.", digest)
-            return True
+    try:
+        # Switch to the target branch by running ``git checkout <branch_name>`` in the target repository.
+        git_obj.repo.git.checkout(res_branch)
+    except GitCommandError as error:
+        logger.error("Cannot checkout branch %s. Error: %s", res_branch, error)
+        return False
 
-        # Check if the commit exist in the history after we have pulled the latest changes.
-        if commit_exists(git_obj, digest):
-            # The digest was found in the history of the branch.
+    logger.info("Successfully checkout branch %s.", res_branch)
+
+    if not offline_mode:
+        # We only pull the latest changes if one of these scenarios happens:
+        #   - no digest is provided: we need to pull and analyze the latest commit.
+        #   - a commit digest is provided but it does not exist locally: we need to
+        #     pull the latest changes to check if that commit is available.
+        # We want to check if the commit already exist locally first because we want to avoid pulling unecessary
+        # if it does.
+        # We do this by checking if the commit we want to analyze is an ancestor of the commit being referenced by HEAD
+        # (which point to the tip of the branch).
+        # If the commit we want to analyze is same as HEAD, that commit is still considered as the ancestor of HEAD.
+        # The ``is_ancestor`` method runs ``git merge-base`` behind the scence.
+        # For more information on computing the ancestor status of two commits: https://git-scm.com/docs/git-merge-base.
+        if not digest or not git_obj.repo.is_ancestor(digest, "HEAD"):
+            logger.info("Pulling the latest changes of branch %s fast-forward only.", res_branch)
             try:
-                git_obj.checkout(digest)
+                # Pull the latest changes on the current branch fast-forward only.
+                git_obj.repo.git.pull("--ff-only")
```

Comment:
Thinking more about it. We don't need to pull before checking out a commit. If there is a newer commit from remote that has been fetched to the local repository, we could checkout that commit directly without pulling. 
In theory, this could save us from downloading the uneccessary trees or blobs of commits that we don't care about. However, I don't have enough evidence to validate it yet. I will continue working it.